src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,356 @@
+using namespace chip::app::Compatibility::Internal;
+
+// Fetch the source for the given attribute path: either a cluster (for global ones) or attribute
+// path.
+//
+// if returning a CHIP_ERROR, it will NEVER be CHIP_NO_ERROR.
+std::variant<const EmberAfCluster *,           // global attribute, data from a cluster
+             const EmberAfAttributeMetadata *, // a specific attribute stored by ember
+             CHIP_ERROR                        // error, this will NEVER be CHIP_NO_ERROR
+             >
+FindAttributeMetadata(const ConcreteAttributePath & aPath)
+{
+    for (auto & attr : GlobalAttributesNotInMetadata)
+    {
+        if (attr == aPath.mAttributeId)
+        {
+            const EmberAfCluster * cluster = emberAfFindServerCluster(aPath.mEndpointId, aPath.mClusterId);
+            ReturnErrorCodeIf(cluster == nullptr, CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute));
+            return cluster;
+        }
+    }
+
+    const EmberAfAttributeMetadata * metadata =
+        emberAfLocateAttributeMetadata(aPath.mEndpointId, aPath.mClusterId, aPath.mAttributeId);
+
+    if (metadata == nullptr)
+    {
+        const EmberAfEndpointType * type = emberAfFindEndpointType(aPath.mEndpointId);
+        if (type == nullptr)
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedEndpoint);
+        }
+
+        const EmberAfCluster * cluster = emberAfFindClusterInType(type, aPath.mClusterId, CLUSTER_MASK_SERVER);
+        if (cluster == nullptr)
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedCluster);
+        }
+
+        // Since we know the attribute is unsupported and the endpoint/cluster are
+        // OK, this is the only option left.
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+
+    return metadata;
+}
+
+/// Attempts to read via an attribute access interface (AAI)
+///
+/// If it returns a CHIP_ERROR, then this is a FINAL result (i.e. either failure or success):
+///    - in particular, CHIP_ERROR_ACCESS_DENIED will be used for UnsupportedRead AII returns
+///
+/// If it returns std::nullopt, then there is no AAI to handle the given path
+/// and processing should figure out the value otherwise (generally from other ember data)
+std::optional<CHIP_ERROR> TryReadViaAccessInterface(const ConcreteAttributePath & path, AttributeAccessInterface * aai,
+                                                    AttributeValueEncoder & encoder)
+{
+    // Processing can happen only if an attribute access interface actually exists..
+    if (aai == nullptr)
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err = aai->Read(path, encoder);
+
+    // explict translate UnsupportedRead to Access denied. This is to allow callers to determine a
+    // translation for this: usually wildcard subscriptions MAY just ignore these where as direct reads
+    // MUST translate them to UnsupportedAccess
+    ReturnErrorCodeIf(err == CHIP_IM_GLOBAL_STATUS(UnsupportedRead), CHIP_ERROR_ACCESS_DENIED);
```

Comment:
This was because of ember-compatibiltiy treating this as access denied:

https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L204

```cpp
    if (err == CHIP_IM_GLOBAL_STATUS(UnsupportedRead) && aPath.mExpanded)
    {
        //
        // Set this to true to ensure our caller will return immediately without proceeding further.
        //
        *aTriedEncode = true;
        return CHIP_NO_ERROR;
    }
```

Unfortunately the comment does not exactly explain why an unsupported read while path-expanding is supposed to be a "success/continue" and the only current expansion logic that I have seen was that "access denied is not an error for path expansion" hence my change.

It may also be that this is for "write-only" attribute skipping, but that seems also to translate reasonably as access denied.

The change is mainly as a "do whatever ember compatibility does".

Reply:
Alternative would be to allow `UnsupportedRead` to pass through and document that as a "potentially skippable for wildcard" as well, however I was unsure.

Thinking about it, I imagine we have certification tests that ask for 'unsupported read needs to be returned' so it seems I should pass it through. I will update the code.

Reply:
Code updated.

Reply:
It's actually reasonably difficult to test unsupported read, unfortunately - we don't have any in the spec, I think we have one as a part of unit testing, but I've only ever encountered them for real in MEI clusters.